src/Toml/Parser/TOML.hs

```diff
@@ -1,44 +1,47 @@
 module Toml.Parser.TOML
        ( keyValP
        , tableHeaderP
+       , inlineTableP
        , tomlP
        ) where
 
-import Control.Applicative (Alternative (many))
+import Control.Applicative (Alternative (many, (<|>)))
+import Control.Applicative.Combinators (sepEndBy, between)
+import Data.Either (lefts, rights)
 
-import Toml.Parser.Core (Parser, sc, text)
-import Toml.Parser.Value (keyP, tableNameP, valueP)
+import Toml.Parser.Core (Parser, sc, text, char, try)
+import Toml.Parser.Value (keyP, tableNameP, anyValueP)
 import Toml.PrefixTree (Key (..), fromList)
-import Toml.Type (AnyValue, TOML (..), typeCheck)
+import Toml.Type (AnyValue, TOML (..))
 
 import qualified Data.HashMap.Lazy as HashMap
 
+hasKeyP :: Parser (Either (Key, AnyValue) (Key, TOML))
+hasKeyP = try (Left <$> keyValP) <|> (Right <$> inlineTableP)
 
 keyValP :: Parser (Key, AnyValue)
-keyValP = do
-    k    <- keyP
-    _    <- text "="
-    uval <- valueP
-    case typeCheck uval of
-        Left err -> fail $ show err
-        Right v  -> pure (k, v)
+keyValP = (,) <$> keyP <* text "=" <*> anyValueP
 
+inlineTableP :: Parser (Key, TOML)
+inlineTableP =
+  (,) <$> keyP <* text "="
+      <*> between (chSc '{') (chSc '}')
+                  (makeToml <$> hasKeyP `sepEndBy` chSc ',')
+  where
+    chSc :: Char -> Parser ()
+    chSc c = char c *> sc
 
 tableHeaderP :: Parser (Key, TOML)
-tableHeaderP = do
-    k    <- tableNameP
-    toml <- makeToml <$> many keyValP
-    pure (k, toml)
-  where
-    makeToml :: [(Key, AnyValue)] -> TOML
-    makeToml kv = TOML (HashMap.fromList kv) mempty
+tableHeaderP = (,) <$> tableNameP <*> (makeToml <$> many hasKeyP)
 
+makeToml :: [Either (Key, AnyValue) (Key, TOML)] -> TOML
+makeToml kvs = TOML (HashMap.fromList $ lefts kvs) (fromList $ rights kvs)
 
 tomlP :: Parser TOML
 tomlP = do
     sc
-    kvs    <- many keyValP
+    kvs    <- many hasKeyP
     tables <- many tableHeaderP
-    pure TOML { tomlPairs  = HashMap.fromList kvs
-              , tomlTables = fromList tables
-              }
+    return TOML { tomlPairs  = HashMap.fromList $ lefts kvs
+                , tomlTables = fromList $ tables ++ rights kvs
```

Comment:
You use `lefts` and `rights` function and because of that you traverse list twice and this is not efficient. It's better if you write polymorphic function like:

```haskell
distributeEithers :: [(c, Either a b)] -> ([(c, a)], [(c, b)])
```